div with rounding modes [+ rounded division]

Keno opened the pull request:
I found myself needing a rounding integer division, i.e. one that does RoundToNearestTiesAway (actually the reference implementation calls for RoundToNearestTiesToward, but the paper doesn't specified and either should be ok). Rather than adding a rntad (round-to-neearest-ties-away-division) function, I figured it may make sense to instead create a generic `div(x, y, r::RoundingMode=RoundToZero)` function, while keeping `fld`,`cld` (and the two-argument div) as useful shorthands for the rounding modes they represent.

~~This PR just does the re-arranging for the existing code. I haven't yet added the new functionality for round-nearest. Let's discuss whether this direction is sensible.~~

Keno committed:
Re-arrange fld/cld code

In preparation for supporting other rounding modes in div, create
a three-argument div function that takes a rounding mode as the last
argument and make this the fundamental fallback for fld/cld.

Keno commented:
Somehow I missed that we already have `rem(x, y, r::RoundingMode)` with compatible semantics, which I thinks validates the sensibility here.

StefanKarpinski commented:
I feel like @simonbyrne would have the most well-informed opinion on this matter but it seems like a solid idea to me?

simonbyrne commented:
:+1: this was my plan from #9283.

Keno committed:
Implemented rounded division

simonbyrne commented on base/div.jl:
```diff
@@ -0,0 +1,188 @@
+"""
+    div(x, y, r::RoundingMode=RoundToZero)
+
+    Compute the remainder of `x` after integer division by `y`, with the quotient rounded
```
unindent this text

simonbyrne commented on base/div.jl:
```diff
@@ -0,0 +1,188 @@
+        return q + copysign(1, q)
+    else
+        @assert rnd == RoundNearestTiesUp
+        return sign(q) == -1 ? q : q + 1
```
I don't think it handles negative odd values of `y` correctly, e.g. `divrem(5, -3, RoundNearestTiesUp)`.

simonbyrne commented on base/div.jl:
```diff
@@ -0,0 +1,188 @@
+                                              typeof(RoundNearestTiesAway),
+                                              typeof(RoundNearestTiesUp)})
+    (q, r) = divrem(x, y)
+    halfy = copysign(y >> 1, r)
```
would be good to explain the logic of this function: I don't quite understand it myself.

simonbyrne commented on test/int.jl:
```diff
@@ -305,3 +305,18 @@ let i = MyInt26779(1)
+            (3, 2, 2, 2, 2),
+            (-3, 2, -2, -2, -1),
+            (5, 2, 2, 3, 3),
+            (-5, 2, -2, -3, -2))
```
Add some odd divisor cases to the test cases. Also `typemin`/`typemax` divisors/dividends.

simonbyrne commented on base/div.jl:
```diff
@@ -0,0 +1,188 @@
+    c == -1 && return q
+    c == 1 && return q + copysign(1, q)
+    if rnd == RoundNearest
+        return iseven(q) ? q : q + copysign(1, q)
```
None of these handle `y == 1` correctly.

simonbyrne commented on base/div.jl:
```diff
@@ -0,0 +1,218 @@
+        # The divisior is odd - no ties are possible, just
+        # round to nearest. N.B. 2r == y is impossible because
+        # y is odd.
+        return abs(2r) > abs(y) ? q + copysign(one(q), q) : q
```
This won't work if `2r` overflows, e.g. `x,y = typemax(Int)-2,typemax(Int)`

simonbyrne committed:
add various divrem combinations to avoid overflow

simonbyrne commented:
@keno I've added some code that I think should handle overflow correctly. Tests could be improved, and I'm not really sure they will handle mixed Int/UInt arguments correctly.

Keno commented:
Oh, I was just gonna use checked_mul, but I like that your version has divrem. We could switch all the divisions to bitshifts, or maybe we just trust that the compiler will figure it out (and that gmp has sufficient special cases). We should also definitely add your overflow test case.

Keno committed:
Whitespace/test fixes

smldis commented on base/div.jl:
```diff
@@ -0,0 +1,244 @@
+div(a, b) = div(a, b, RoundToZero)
+
+"""
+    rem(x, y, r::RoundingMode)
```
maybe `rem(x, y, r::RoundingMode=RoundToZero)` as div docstring

Keno committed:
Small tweaks to docstrings

simonbyrne commented:
ooh, using `checked_mul` is a good idea. It should be possible to rewrite my code to use that instead.

Keno commented:
(Just as a note, I meant mul_with_overflow, the checked_ one throws).

Keno committed:
Bugfixes

Keno commented:
Rebased and fixed a few more issues. If CI passes, I'd like to start getting this in.

simonbyrne commented:
If we want to check for overflow, we could exhaustively test `Int8`.

Keno committed:
Add the exhaustive test

Keno commented:
> If we want to check for overflow, we could exhaustively test Int8.

Good idead. Done. This should be good to go from my side.

Keno committed:
Tigthen up types for div fallback

I think it's better to give a MethodError here than an approximate
answer for non-AbstractFloat reals (e.g. custom integer types).

NHDaly commented:
I noticed that `fld1`  isn't included in this change, and it looks like it's still calling two-argument `div`:
https://github.com/JuliaLang/julia/blob/09d617a7db4002bd499d0cf3894ccf2d2cd23b33/base/operators.jl#L770

Should packages still implement a custom override for `fld1` if needed? (I'm looking at adding this support for FixedDecimals, and i'm wondering if this can be as simple as just adding `Base.div(x::T, y::T, r::RoundingMode) where {T <: FD} = T(div(x.i, y.i, r))`? Do I need to stop defining two-arg-`div`, `fld`, `fld1`, etc?
